user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java

```diff
@@ -139,12 +151,127 @@ protected void createOrUpdate(Reconciliation reconciliation, KafkaUser kafkaUser
-                .map((Void) null).setHandler(handler);
+                .compose(reconciliationResult -> {
+                    Condition readyCondition;
+                    if (kafkaUser.getMetadata().getGeneration() != null)    {
+                        userStatus.setObservedGeneration(kafkaUser.getMetadata().getGeneration());
+                    }
+
+                    if (reconciliationResult.succeeded()) {
+                        readyCondition = new ConditionBuilder()
+                                .withNewLastTransitionTime(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(dateSupplier()))
+                                .withNewType("Ready")
+                                .withNewStatus("True")
+                                .build();
+                    } else {
+                        readyCondition = new ConditionBuilder()
+                                .withNewLastTransitionTime(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(dateSupplier()))
+                                .withNewType("NotReady")
+                                .withNewStatus("True")
+                                .withNewReason(reconciliationResult.cause().getClass().getSimpleName())
+                                .withNewMessage(reconciliationResult.cause().getMessage())
+                                .build();
+                    }
+
+                    userStatus.setUsername(user.getName());
+                    userStatus.setConditions(Collections.singletonList(readyCondition));
+
+                    updateStatus(kafkaUser, reconciliation, userStatus).setHandler(statusResult -> {
+                        if (statusResult.succeeded())    {
+                            log.debug("Status for {} is up to date", kafkaUser.getMetadata().getName());
+                        } else {
+                            log.error("Failed to set status for {}. {}", kafkaUser.getMetadata().getName(), statusResult.cause().getMessage());
+                        }
+
+                        // If both features succeeded, createOrUpdate succeeded as well
+                        // If one or both of them failed, we prefer the reconciliation failure as the main error
+                        if (reconciliationResult.succeeded() && statusResult.succeeded())    {
+                            createOrUpdateFuture.complete();
+                        } else if (reconciliationResult.failed())    {
+                            createOrUpdateFuture.fail(reconciliationResult.cause());
+                        } else {
+                            createOrUpdateFuture.fail(statusResult.cause());
+                        }
+                    });
+                    return createOrUpdateFuture;
+                }).map((Void) null).setHandler(handler);
+    }
+
+    private Future<ReconcileResult<Secret>> setSecretStatus(String namespace, KafkaUserModel user, Secret desired, KafkaUserStatus userStatus) {
```

Comment:
I think calling this `setSecretStatus` is misleading. The main thing is the secret reconciliation.